src/carriers/mpi_carrier/include/yarp/os/MpiBcastCarrier.h

```diff
@@ -68,7 +68,8 @@ class yarp::os::MpiBcastCarrier : public MpiCarrier {
 
     virtual bool expectReplyToHeader(ConnectionState&  proto) override {
         bool ok = MpiCarrier::expectReplyToHeader(proto);
-        dynamic_cast<MpiBcastStream*> (stream)->post();
+        if(dynamic_cast<MpiBcastStream*> (stream))
+            dynamic_cast<MpiBcastStream*> (stream)->post();
```

Comment:
the dynamic_cast twice, and it is useless to call it twice here, please save the pointer instead

Reply:
Btw, not directly related to the PR here, but a `dynamic_cast` is quite costly, and this one is totally useless and in hot path, as:
* we are not checking the output of the `dynamic_cast`, i.e. if it is null or not
* we explicitly creating the stream pointer as  `MpiBcastStream`, so we now that it will be pointing to an object of type MpiBcastStream`. 
I think it can be safely removed we can write an issue for that.

Reply:
I agree. Also because I think that `expectReplyToHeader` is called for every message.

Anyway I think that removing this `dynamic_cast` this is outside the scope of this PR, so I'd rather just fix it to save the pointer here, and eventually fix this later...

Reply:
Fixed https://github.com/robotology/yarp/pull/1412/commits/c6bf0e8d555ed4134e56db7bcf237c5bc4c153ef